src/select/SelectMenu.ts

```diff
@@ -1,77 +1,56 @@
-import { createHook, createComponent } from "reakit-system";
+import { SelectStateReturn } from "./SelectState";
 import { SELECT_KEYS } from "./__keys";
-
-export type SelectMenuOptions = Pick<
-  SelectStateReturn,
-  "setTypehead" | "selected" | "hide" | "show" | "visible"
-> & {
-  onChange?: (value: any) => void;
-};
+import { createComponent, createHook } from "reakit-system";
+import { CompositeProps, useComposite } from "reakit/Composite";
+import { usePortalShortcut } from "./__utils";
+import { useForkRef } from "reakit-utils";
+
+export type SelectMenuOptions = CompositeProps &
+  Pick<
+    SelectStateReturn,
+    "setSelected" | "visible" | "values" | "currentId" | "move"
+  > & {
+    maxHeight?: string | number;
+    modal?: boolean;
+  };
 
 const useSelectMenu = createHook<SelectMenuOptions, BoxHTMLProps>({
   name: "SelectMenu",
-  keys: SELECT_KEYS,
+  compose: [useComposite, usePopover],
+  keys: ["maxHeight", ...SELECT_KEYS],
+  useOptions({ modal = false, ...options }) {
+    return { modal, ...options };
+  },
 
   useProps(
-    { setTypehead, selected, onChange, hide, show, visible },
-    { ...htmlProps },
+    { maxHeight = 500, visible, move, values, currentId },
+    { ref: htmlRef, style: htmlStyle, ...htmlProps },
   ) {
-    const keyClear = React.useRef<any>(null);
-    const [typed, setTyped] = React.useState("");
-
-    React.useEffect(() => {
-      onChange && onChange(selected);
-    }, [selected]);
-
-    const onKeyDown = React.useMemo(() => {
-      return createOnKeyDown({
-        stopPropagation: true,
-        keyMap: () => ({
-          Escape: hide,
-          ArrowUp: show,
-          ArrowDown: show,
-        }),
-      });
-    }, []);
+    const ref = React.useRef<HTMLElement>(null);
 
-    const clearKeyStrokes = () => {
-      setTypehead(typed);
-
-      if (keyClear.current) {
-        clearTimeout(keyClear.current);
-        keyClear.current = null;
-      }
-
-      keyClear.current = setTimeout(() => {
-        setTyped("");
-        keyClear.current = null;
-      }, 800);
-    };
+    usePortalShortcut(ref, { values, currentId, move });
 
     React.useEffect(() => {
-      if (typed !== "") {
-        clearKeyStrokes();
+      if (values) {
+        const firstItem = values[0];
+        firstItem && move?.(firstItem.id);
       }
-    }, [typed]);
-
-    const handleOnKeyPress = (e: React.KeyboardEvent) => {
-      e.persist();
-      // skip the enter key
-      if (e.key === "Enter") return;
-      setTyped(prev => prev + e.key);
-    };
+    }, [visible]);
 
     return {
-      role: "button",
-      "aria-expanded": visible,
-      "aria-haspopup": "listbox",
-      onKeyDown,
-      onKeyPress: handleOnKeyPress,
+      ref: useForkRef(ref, htmlRef),
+      role: "listbox",
+      "aria-orientation": "vertical",
+      "aria-hidden": !visible,
+      style: {
+        maxHeight: maxHeight,
+        overflowY: "scroll",
+        ...htmlStyle,
+      },
```

Comment:
Do you think, we need to set these styles.
Instead of the user handling this?
If they want a scrolling container, they can set an maxHeight and the overflowY themselves right !!

Reply:
Okay will remove it. then we also need the maxHeight option